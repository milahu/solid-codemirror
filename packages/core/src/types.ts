export interface CodeMirrorProps {
  value?: string;
  onValueChange?: (value: string) => void;
}
