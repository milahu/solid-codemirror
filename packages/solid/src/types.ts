export interface CodeMirrorProps {
  value?: string;
  onChange?: (value: string) => void;
}
