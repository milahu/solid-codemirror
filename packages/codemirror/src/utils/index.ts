export * from "./createLineNumbers";
export * from "./createReadOnly";
