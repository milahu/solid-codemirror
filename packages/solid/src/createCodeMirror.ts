import { Accessor, onCleanup, onMount, on, createEffect } from "solid-js";
import { EditorView } from "@codemirror/view";
import {
  Compartment,
  EditorState,
  Extension,
  StateEffect,
} from "@codemirror/state";
import { CodeMirrorProps } from "./types";

export function createCodeMirror(
  props: CodeMirrorProps,
  ref: Accessor<HTMLDivElement | undefined>
) {
  let view: EditorView;

  onMount(() => {
    const state = EditorState.create({
      doc: props.value,
    });

    view = new EditorView({
      state,
      parent: ref(),
      dispatch: (tr): void => {
        view.update([tr]);

        if (tr.docChanged) {
          const newCode = tr.newDoc.sliceString(0, tr.newDoc.length);
          props.onValueChange?.(newCode);
        }
      },
    });

    onCleanup(() => {
      view.destroy();
    });
  });

  createEffect(
    on(
      () => props.value,
      (value) => {
        if (value === view.state.doc.toString()) {
          return;
        }
        view.dispatch({
          changes: {
            from: 0,
            to: view.state.doc.length,
            insert: value,
          },
        });
      },
      { defer: true }
    )
  );

  function createExtension(extension: Extension) {
    const compartment = new Compartment();

    onMount(() => {
      view.dispatch({
        effects: StateEffect.appendConfig.of(compartment.of(extension)),
      });
    });

    function reconfigure(extension: Extension) {
      view.dispatch({
        effects: compartment.reconfigure(extension),
      });
    }

    return reconfigure;
  }

  return { createExtension };
}
