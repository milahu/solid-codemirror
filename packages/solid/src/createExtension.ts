import { Compartment, Extension, StateEffect } from "@codemirror/state";
import { EditorView } from "@codemirror/view";
import { Accessor, createEffect, on } from "solid-js";

export function createExtension<T>(view: Accessor<EditorView | undefined>) {
  const compartment = new Compartment();

  createEffect(
    on(
      view,
      (view) => {
        console.log(view);
        if (view) {
          view.dispatch({
            effects: StateEffect.appendConfig.of(compartment.of([])),
          });
        }
      },
      { defer: true }
    )
  );

  createEffect(
    on(
      [
        /* dependency */
      ],
      (value) => {
        console.log(view());
        view()?.dispatch({
          effects: compartment.reconfigure([]),
        });
      },
      { defer: true }
    )
  );
}
