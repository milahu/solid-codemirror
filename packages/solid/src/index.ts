export * from "./createCodeMirror";
export * from "./types";
