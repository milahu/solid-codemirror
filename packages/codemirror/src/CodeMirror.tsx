import { splitProps } from "solid-js";
import type { JSX } from "solid-js/jsx-runtime";
import { mergeRefs } from "@solid-primitives/refs";
import { type CodeMirrorProps, createCodeMirror  } from "@solid-codemirror/core";
import { createLineNumbers, createReadOnly, LineNumbersProps, ReadOnlyProps } from "./utils";

type Props = CodeMirrorProps & LineNumbersProps & ReadOnlyProps & JSX.HTMLAttributes<HTMLDivElement>

export function CodeMirror(
  props: Props
) {
  let ref: HTMLDivElement | undefined;

  const [codemirrorProps, lineNumberProps, readOnlyProps, others] = splitProps(props, ["value", "onValueChange"], ["showLineNumbers"], ["readOnly"]);
  const { createExtension } = createCodeMirror(codemirrorProps, () => ref);

  createLineNumbers(lineNumberProps, createExtension);
  createReadOnly(readOnlyProps, createExtension);

  return <div ref={mergeRefs(el => (ref = el), props.ref)} {...others} />;
}


