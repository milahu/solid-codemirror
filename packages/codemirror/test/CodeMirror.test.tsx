import { describe, test, expect, it, vi, afterEach } from "vitest";
import { fireEvent, screen, render } from "solid-testing-library";
import { createRoot } from "solid-js";
import { CodeMirror } from "../src";

describe("createCodeMirror", () => {
  it("renders", async () => {
    const { container, unmount } = render(() => <CodeMirror />);
    expect(container.innerHTML).toMatchSnapshot();
    unmount();
  });
});
