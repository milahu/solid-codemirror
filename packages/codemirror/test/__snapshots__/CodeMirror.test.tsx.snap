// Vitest Snapshot v1

exports[`createCodeMirror > renders 1`] = `"<div></div>"`;
